ElasticSearch Query Editor - Scripted Metric

shanielh opened the pull request:
Further to #3441 

A pull request to add `Scripted Metric` option to query builder.

Couldn't do the "custom aggregation", too hard for me, Would love to get some guidance from someone on that one.

![scripted metric](https://cloud.githubusercontent.com/assets/746471/11615311/cfd095e0-9c65-11e5-891a-92b56fb5d46a.png)

shanielh committed:
Added Scripted Metric

shanielh commented on public/app/plugins/datasource/elasticsearch/partials/metricAgg.html:
```diff
@@ -62,5 +65,20 @@
 		</div>
+
+		<div ng-if="agg.type === 'scripted_metric'">
+			<div class="tight-form" ng-repeat="option in scriptedMetricOptions" ng-class="{last: $last}">
```
Can I use here `aggDef`?

torkelo commented:
this looks like an experimental Elasticsearch feature so it might have to wait before being included in Grafana

shanielh commented:
This feature is available since elastic 1.4 (https://www.elastic.co/guide/en/elasticsearch/reference/1.4/search-aggregations-metrics-scripted-metric-aggregation.html). It's still considered experimental, but it's old feature. Maybe we can work on solution for "advanced" users only?

Qtax commented:
Maybe we could also use the proposed ES datasource "enable scripts" setting in https://github.com/grafana/grafana/pull/3438 for this too? In which case we could hide this unless the option is enabled (disabled by default).

Qtax commented:
Imho it could be nice to see a `lang` setting. (In case you don't want to change the default on ES.)
Don't know if ppl really need it, but I would like to see it somewhere in Grafana at least. If not desired on every metric, then possibly on the datasource.

shanielh commented:
@Qtax :+1: I also need the scripting abilities because 2.6 disabled the manual query editor.

shanielh committed:
Scripted Metric marked as Experimental Feature

It will be shown only when the user chooses to allow experimental features
in the elastic-search datasource configuration panel.

shanielh commented:
@Qtax @torkelo Hi, I haven't added the language feature because I figured out from the other PR (#3438) that @torkelo doesn't want this setting.

But I did added the "Experimental Features" option in the ElasticSearch Datasource configuration panel : 

![image](https://cloud.githubusercontent.com/assets/746471/11762297/9fc44232-a0ea-11e5-9252-bebedf032984.png)

The "Scripted" Metric feature will be shown only when that option is checked.

I hope that after this edit, This pull request will be merged. @torkelo - If there is anything you want me to change, Please tell me.

barvinograd commented:
+1

shanielh commented:
@torkelo Any update regarding this? It will be helpful for me. 

I cannot updated to 2.6 because `Editor Mode` is not supported in that version and I'm looking forward to upgrade for other bugs fixes.

Elastic 2.2 is out and scripted metric is [still](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-scripted-metric-aggregation.html) supported.

ozarvatz commented:
:+1:

eliranmoyal commented:
+1

OrenDor commented:
:+1:

jisaitua commented:
+1

torkelo commented:
closing this as this PR has drifted a bit too far from master (many conflicts). and needs a bit more reactions / love / demand on the feature request issue: https://github.com/grafana/grafana/issues/3441
